### ISSUE TITLE
Negative loss encountered during BC Transformer training with GMM in RoboCasa

### ISSUE THREAD

**Issue by SII-ZhangYiFei** [robocasa/robocasa#163]
### **Describe the bug**
When training a BC Transformer with GMM enabled on a RoboCasa task, the training loss becomes negative.
The same setup works normally when GMM is disabled, and loss values remain positive and stable.

I am using the GR00T dataset ([link](https://huggingface.co/datasets/GR00T-Dateset/single_panda_gripper-TurnSinkSpout)), processed to match the RoboCasa dataset format (100 demos). Training proceeds correctly with a standard BC Transformer, but enabling GMM causes negative losses to appear after several iterations.

### **Config**
```json
{
    "algo_name": "bc",
    "experiment": {
        "name": "seed_123_ds_CloseDoubleDoor",
        "validate": false,
        "logging": {
            "terminal_output_to_txt": true,
            "log_tb": true,
            "log_wandb": false,
            "wandb_proj_name": "bc_xfmr_10-13-None_robocasa_im"
        },
        "mse": {
            "enabled": false,
            "every_n_epochs": 50,
            "on_save_ckpt": true,
            "num_samples": 20,
            "visualize": true
        },
        "save": {
            "enabled": true,
            "every_n_seconds": null,
            "every_n_epochs": 20,
            "epochs": [],
            "on_best_validation": false,
            "on_best_rollout_return": false,
            "on_best_rollout_success_rate": false
        },
        "epoch_every_n_steps": 671,
        "validation_epoch_every_n_steps": 10,
        "env": null,
        "additional_envs": null,
        "render": false,
        "render_video": true,
        "keep_all_videos": true,
        "video_skip": 5,
        "rollout": {
            "enabled": true,
            "n": 100,
            "horizon": 630,
            "rate": 20,
            "warmstart": 40,
            "terminate_on_success": true,
            "batched": true,
            "num_batch_envs": 20
        },
        "env_meta_update_dict": {
            "env_kwargs": {
                "generative_textures": null,
                "scene_split": null,
                "style_ids": null,
                "layout_ids": null,
                "layout_and_style_ids": [
                    [
                        1,
                        1
                    ],
                    [
                        2,
                        2
                    ],
                    [
                        4,
                        4
                    ],
                    [
                        6,
                        9
                    ],
                    [
                        7,
                        10
                    ]
                ],
                "randomize_cameras": false,
                "obj_instance_split": "B"
            }
        },
        "ckpt_path": null
    },
    "train": {
        "data": [
            {
                "path": path_to_dataset,
                "horizon": 430,
                "do_eval": true,
                "filter_key": null,
                "lang": "close the cabinet doors"
            }
        ],
        "output_dir": output_dir
        "num_data_workers": 24,
        "hdf5_cache_mode": "low_dim",
        "hdf5_use_swmr": true,
        "hdf5_load_next_obs": false,
        "hdf5_normalize_obs": false,
        "hdf5_filter_key": null,
        "hdf5_validation_filter_key": null,
        "seq_length": 10,
        "pad_seq_length": true,
        "frame_stack": 10,
        "pad_frame_stack": true,
        "dataset_keys": [
            "action_dict/rel_pos",
            "action_dict/rel_rot_axis_angle",
            "action_dict/gripper",
            "action_dict/base_mode"
        ],
        "action_keys": [
            "action_dict/rel_pos",
            "action_dict/rel_rot_axis_angle",
            "action_dict/gripper",
            "action_dict/base_mode"
        ],
        "action_config": {
            "actions": {
                "normalization": null
            },
            "actions_abs": {
                "normalization": "min_max"
            },
            "action_dict/abs_pos": {
                "normalization": "min_max"
            },
            "action_dict/abs_rot_axis_angle": {
                "normalization": "min_max",
                "format": "rot_axis_angle"
            },
            "action_dict/abs_rot_6d": {
                "normalization": null,
                "format": "rot_6d"
            },
            "action_dict/rel_pos": {
                "normalization": null
            },
            "action_dict/rel_rot_axis_angle": {
                "normalization": null,
                "format": "rot_axis_angle"
            },
            "action_dict/rel_rot_6d": {
                "normalization": null,
                "format": "rot_6d"
            },
            "action_dict/gripper": {
                "normalization": null
            },
            "action_dict/base_mode": {
                "normalization": null
            }
        },
        "goal_mode": null,
        "cuda": true,
        "batch_size": 64,
        "num_epochs": 200,
        "seed": 123,
        "max_grad_norm": 100.0,
        "data_format": "robomimic",
        "shuffled_obs_key_groups": null
    },
    "algo": {
        "optim_params": {
            "policy": {
                "optimizer_type": "adamw",
                "learning_rate": {
                    "initial": 0.0001,
                    "decay_factor": 1.0,
                    "epoch_schedule": [
                        20
                    ],
                    "scheduler_type": "constant_with_warmup"
                },
                "regularization": {
                    "L2": 0.01
                }
            }
        },
        "loss": {
            "l2_weight": 1.0,
            "l1_weight": 0.0,
            "cos_weight": 0.0
        },
        "actor_layer_dims": [],
        "gaussian": {
            "enabled": false,
            "fixed_std": false,
            "init_std": 0.1,
            "min_std": 0.01,
            "std_activation": "softplus",
            "low_noise_eval": true
        },
        "gmm": {
            "enabled": true,
            "num_modes": 5,
            "min_std": 0.005,
            "std_activation": "softplus",
            "low_noise_eval": true
        },
        "vae": {
            "enabled": false,
            "latent_dim": 14,
            "latent_clip": null,
            "kl_weight": 1.0,
            "decoder": {
                "is_conditioned": true,
                "reconstruction_sum_across_elements": false
            },
            "prior": {
                "learn": false,
                "is_conditioned": false,
                "use_gmm": false,
                "gmm_num_modes": 10,
                "gmm_learn_weights": false,
                "use_categorical": false,
                "categorical_dim": 10,
                "categorical_gumbel_softmax_hard": false,
                "categorical_init_temp": 1.0,
                "categorical_temp_anneal_step": 0.001,
                "categorical_min_temp": 0.3
            },
            "encoder_layer_dims": [
                300,
                400
            ],
            "decoder_layer_dims": [
                300,
                400
            ],
            "prior_layer_dims": [
                300,
                400
            ]
        },
        "rnn": {
            "enabled": false,
            "horizon": 10,
            "hidden_dim": 400,
            "rnn_type": "LSTM",
            "num_layers": 2,
            "open_loop": false,
            "kwargs": {
                "bidirectional": false
            }
        },
        "transformer": {
            "enabled": true,
            "context_length": 10,
            "embed_dim": 512,
            "num_layers": 6,
            "num_heads": 8,
            "emb_dropout": 0.1,
            "attn_dropout": 0.1,
            "block_output_dropout": 0.1,
            "sinusoidal_embedding": false,
            "activation": "gelu",
            "supervise_all_steps": true,
            "nn_parameter_for_timesteps": true,
            "pred_future_acs": true,
            "causal": false
        },
        "language_conditioned": false
    },
    "observation": {
        "modalities": {
            "obs": {
                "low_dim": [
                    "robot0_base_to_eef_pos",
                    "robot0_base_to_eef_quat",
                    "robot0_base_pos",
                    "robot0_base_quat",
                    "robot0_gripper_qpos"
                ],
                "rgb": [
                    "robot0_agentview_left_image",
                    "robot0_agentview_right_image",
                    "robot0_eye_in_hand_image"
                ],
                "depth": [],
                "scan": []
            },
            "goal": {
                "low_dim": [],
                "rgb": [],
                "depth": [],
                "scan": []
            }
        },
        "encoder": {
            "low_dim": {
                "core_class": null,
                "core_kwargs": {},
                "obs_randomizer_class": null,
                "obs_randomizer_kwargs": {}
            },
            "rgb": {
                "core_class": "VisualCoreLanguageConditioned",
                "core_kwargs": {
                    "feature_dimension": 64,
                    "backbone_class": "ResNet18ConvFiLM",
                    "backbone_kwargs": {
                        "pretrained": false,
                        "input_coord_conv": false
                    },
                    "pool_class": "SpatialSoftmax",
                    "pool_kwargs": {
                        "num_kp": 32,
                        "learnable_temperature": false,
                        "temperature": 1.0,
                        "noise_std": 0.0
                    }
                },
                "obs_randomizer_class": "CropRandomizer",
                "obs_randomizer_kwargs": {
                    "crop_height": 76,
                    "crop_width": 76,
                    "num_crops": 1,
                    "pos_enc": false
                }
            },
            "depth": {
                "core_class": "VisualCore",
                "core_kwargs": {},
                "obs_randomizer_class": null,
                "obs_randomizer_kwargs": {}
            },
            "scan": {
                "core_class": "ScanCore",
                "core_kwargs": {},
                "obs_randomizer_class": null,
                "obs_randomizer_kwargs": {}
            }
        }
    },
    "meta": {
        "hp_base_config_file":  path_to_dataset,
        "hp_keys": [
            "seed",
            "ds",
            "obsrandargs"
        ],
        "hp_values": [
            123,
            "CloseDoubleDoor",
            {
                "crop_height": 76,
                "crop_width": 76,
                "num_crops": 1,
                "pos_enc": false
            }
        ]
    }
}
```

### **Train log**
When i set GMM enabled to False, it output:
```
Train Epoch 80
{
    "Cosine_Loss": 0.023647942235374104,
    "L1_Loss": 0.0012381650508664572,
    "L2_Loss": 0.0024772948499050007,
    "Loss": 0.0024772948499050007,
    "Optimizer/policy0_lr": 0.00010000000000000003,
    "Policy_Grad_Norms": 0.004224118778093249,
    "Time_Data_Loading": 1.7248517632484437,
    "Time_Epoch": 17.28038469552994,
    "Time_Log_Info": 0.001128224531809489,
    "Time_Process_Batch": 12.12323574622472,
    "Time_Train_Batch": 3.3410221497217814
}
```
and the rollout is working correctly:
```
Epoch 80 Rollouts took 28.350071625709543s (avg) with results:
Env: CloseDoubleDoor
{
    "Horizon": 411.36,
    "Return": 0.0,
    "Success_Rate": 0.25,
    "Time_Episode": 47.250119376182575,
    "time": 28.350071625709543
}
```

but when i set GMM enabled to true, it turns out a negitive loss and a really huge grad_norm
```
{
    "Log_Likelihood": 0.6615586271735072,
    "Loss": -0.6615586271735072,
    "Optimizer/policy0_lr": 3.36e-06,
    "Policy_Grad_Norms": 3577.8530355282987,
    "Time_Data_Loading": 1.1045984307924905,
    "Time_Epoch": 12.486143616835276,
    "Time_Log_Info": 0.0005890727043151855,
    "Time_Process_Batch": 10.359167794386545,
    "Time_Train_Batch": 1.0023290991783143
}
```

Is this expected behavior or a bug?
If not, what would be the correct way to run the BC Transformer GMM Baseline?
Are there any additional steps I should take when using RoboCasa?

**Comment by SII-ZhangYiFei** [comment]
I processed the dataset like the following:
```python
import os
import h5py
import json
import decord

import pandas as pd
import numpy as np

from tqdm import tqdm

import robomimic.envs.env_base as EB
from robosuite.controllers import load_composite_controller_config

single_panda_tasks = ['CloseDoubleDoor', 'PnPCounterToCab', 'TurnOffMicrowave', 'PnPCounterToMicrowave', 'PnPCounterToSink', 'PnPSinkToCounter', 'PnPCounterToStove', 'TurnOnSinkFaucet', 'OpenSingleDoor', 'PnPMicrowaveToCounter', 'OpenDrawer', 'TurnSinkSpout', 'PnPCabToCounter', 'TurnOffSinkFaucet', 'CloseDrawer', 'TurnOnMicrowave', 'TurnOffStove', 'CoffeeServeMug', 'OpenDoubleDoor', 'CoffeeSetupMug', 'PnPStoveToCounter', 'TurnOnStove', 'CloseSingleDoor', 'CoffeePressButton', 'robocasa_all_tasks']

traj_cnt = 100
# task = "CloseDoubleDoor"
for task in single_panda_tasks:
    base_path = f"{base_path_to_GR00T_Dataset}/single_panda_gripper-{task}"
    episode_info = []
    with open(os.path.join(base_path, "meta/episodes.jsonl"), 'r') as f:
        for line in f:
            episode_info.append(json.loads(line))

    with h5py.File(f"./datasets/{task}.hdf5", "w") as f:
        data_grp = f.create_group("data")
        for i in tqdm(range(int(traj_cnt))):  # demo_0, demo_1, demo_2
            grp = data_grp.create_group(f"demo_{i+1}")
            traj_len = episode_info[i]['length']
            with open(os.path.join(base_path, "meta/modality.json")) as m_f:
                modality = json.load(m_f)
            state_key = {}
            action_key = {}
            video_key = {}
            for view in modality['video'].keys():
                video_key[view] = modality['video'][view]['original_key']
            for joint in modality['action'].keys():
                action_key[joint] = (modality['action'][joint]['start'], modality['action'][joint]['end'])
            for joint in modality['state'].keys():
                state_key[joint] = (modality['state'][joint]['start'], modality['state'][joint]['end'])
            
            videos = {}
            for view in video_key.keys():
                vr = decord.VideoReader(os.path.join(base_path, "videos", f"chunk-{(i//1000):03d}", video_key[view], f"episode_{i:06d}.mp4"))
                frames = vr.get_batch(range(traj_len)).asnumpy()
                videos[view] = frames

            data_file = pd.read_parquet(os.path.join(base_path, "data", f"chunk-{(i//1000):03d}", f"episode_{i:06d}.parquet"))
            state = np.stack(data_file['observation.state'])
            action = np.stack(data_file['action'])
            reward = np.stack(data_file['next.reward'])
            done = np.stack(data_file['next.done'])
            next_state = np.zeros_like(state)
            next_state[:-1] = state[1:]
            next_state[-1] = state[-1]
            next_videos = {}
            for view in videos.keys():
                next_videos[view] = videos[view]
                next_videos[view][:-1] = videos[view][1:]
                next_videos[view][-1] = videos[view][-1]
            
            # # Add datasets to each demo
            ac_grp = grp.create_group("action_dict")
            obs_grp = grp.create_group("obs")
            next_obs_grp = grp.create_group("next_obs")

            data_key_map = {
                'base_position': 'robot0_base_pos',
                'base_rotation': 'robot0_base_quat',
                'gripper_qpos': 'robot0_gripper_qpos',
                'end_effector_position_relative': 'robot0_base_to_eef_pos',
                'end_effector_rotation_relative': 'robot0_base_to_eef_quat',
                'right_view': 'robot0_agentview_right_image',
                'left_view': 'robot0_agentview_left_image',
                'wrist_view': 'robot0_eye_in_hand_image',
            }
            
            # for key in action_key.keys():
            #     act_grp.create_dataset(data_key_map[key], data=action[:traj_len, action_key[key][0]:action_key[key][1]])
            
            for key in state_key.keys():
                obs_grp.create_dataset(data_key_map[key] if key in data_key_map.keys() else key, data=state[:traj_len, state_key[key][0]:state_key[key][1]])
            
            for key in video_key.keys():
                obs_grp.create_dataset(data_key_map[key] if key in data_key_map.keys() else key, data=videos[key][:traj_len])
            
            for key in state_key.keys():
                next_obs_grp.create_dataset(data_key_map[key] if key in data_key_map.keys() else key, data=next_state[:traj_len, state_key[key][0]:state_key[key][1]])
            
            for key in video_key.keys():
                next_obs_grp.create_dataset(data_key_map[key] if key in data_key_map.keys() else key, data=next_videos[key][:traj_len])
            
            ac_grp.create_dataset("base_mode", data=action[:traj_len, :5])
            ac_grp.create_dataset("rel_pos", data=action[:traj_len, 5:8])
            ac_grp.create_dataset("rel_rot_axis_angle", data=action[:traj_len, 8:11])
            ac_grp.create_dataset("gripper", data=action[:traj_len, 11].reshape(traj_len, 1))
            
            grp.create_dataset("dones", data=done)
            grp.create_dataset("rewards", data=reward)
            grp.create_dataset("states", data=np.zeros((traj_len, 1)))
            grp.attrs['num_samples'] = traj_len
        mask_train = np.array([f"demo_{i+1}" for i in range(traj_cnt)], dtype='|S10')
        mask_grp = f.create_group("mask")
        mask_grp.create_dataset("train", data=mask_train)
        # robosuite-related configs
        robots="PandaOmron"
        camera_names=[
            "robot0_agentview_left",
            "robot0_agentview_right",
            "robot0_eye_in_hand",
        ]
        camera_widths=256
        camera_heights=256
        controller_config = load_composite_controller_config(
            controller=None,
            robot=robots if isinstance(robots, str) else robots[0],
        )
        render_onscreen = False
        env_kwargs = dict(
            env_name=task,
            robots=robots,
            controller_configs=controller_config,
            camera_names=camera_names,
            camera_widths=camera_widths,
            camera_heights=camera_heights,
            has_renderer=render_onscreen,
            has_offscreen_renderer=(not render_onscreen),
            ignore_done=True,
            use_object_obs=True,
            use_camera_obs=(not render_onscreen),
            camera_depths=False,
            seed=42,
            obj_instance_split=None,
            generative_textures=None,
            randomize_cameras=None,
            layout_and_style_ids=None,
            layout_ids=None,
            style_ids=None,
            translucent_robot=False,
        )
        env_meta = dict(
            type=1,  # hardcode as robosuite dataset type
            env_name=task,
            env_kwargs=env_kwargs
        )
        # f["data"].attrs["env_args"] = json.dumps(env_meta, indent=4)
        data_grp.attrs["env_args"] = json.dumps(env_meta, indent=4)
```

**Comment by Abhiram824** [comment]
I believe the loss used is the negative [log likelihood loss](https://docs.pytorch.org/docs/stable/generated/torch.nn.NLLLoss.html) and negative values are okay
